### ISSUE TITLE
Emphasize more strongly that contributors should use pre-commit hooks

### ISSUE THREAD

**Issue by MarkusPiotrowski** [biopython/biopython#2014]
While Biopython is enforcing many/most coding-style conventions described in PEP-8 and PEP-257, many pull requests get delayed, because contributors (often unintentionally) violate these style conventions.

A discussion about a Biopython specific pre-commit git hook to enforce PEP-8 style has already started some time ago (e.g. #493, #1188).

Using a pre-commit hook in the git environment is an easy way to make contributors aware of style-violations and give them the chance to address these before submitting a pull request.  

However, it seems that many contributors are not using pre-commit hooks, and one reason for this may be that they are not aware of them and/because we are not explicit enough about that they *should* be used. 

There are some places, where we could emphasize this more strongly:

- ```README.rst``` should contain, under 'Contributing', a link to ```CONTRIBUTING.rst```.
- ```CONTRIBUTING.rst``` should strongly recommend the use of a pre-commit hook and show how to install and set it up.
- There should be a clear recommendation, which tools should be used. 
- PR #1188 addresses already some of these topics and should be merged. It contains a script that will setup the git environment with our folder-specific ```.flake8``` settings.
- The suggested changes also apply to our Biopython home page: ```GitUsage.md``` (https://biopython.org/wiki/GitUsage). 
- We can also think about putting another (optional) paragraph in our PR template https://github.com/biopython/biopython/blob/master/.github/PULL_REQUEST_TEMPLATE.md or rephrase the second paragraph to mention pre-commit hooks.

**Comment by peterjc** [comment]
This is a very good idea in principle, but the complexities of our different ``.flake8`` settings in different folders is an annoyance requiring the bespoke git commit script.

I would very much like to address that first, which bring the enormous advantage of letting us use an off-the shelf widely used hook, e.g. that provided by ``flake8`` itself:

```
$ flake8 --install-hook git
$ git config --bool flake8.strict true
```

Currently (leaving aside optional extensions we don't currently require):

https://github.com/biopython/biopython/blob/biopython-173/BioSQL/.flake8 - quite strict
- ignores W504,D101,D102,D103,D105,

https://github.com/biopython/biopython/blob/biopython-173/Doc/examples/.flake8 - quite strict
- ignores D100,D101,D102,D103

https://github.com/biopython/biopython/blob/biopython-173/Bio/.flake8 - fairly strict
 - ignores E501,E731,F401,F841,D100,D101,D102,D103,D105,D203,

https://github.com/biopython/biopython/blob/biopython-173/Tests/.flake8 - least strict
 - ignores E402,E501,E731,F401,F405,F841,D100,D101,D102,D103,D105,D200,D202,D203,D204,    D300,D400,D401,D412

This looks doable (and is part of my motivation with adopting the black formatting tool, #2008, which would take care of some of these).

Basically we need to fix a few things in the tests and/or relax the other tests a little.

**Comment by peterjc** [comment]
I'm looking at D300 right now...

**Comment by peterjc** [comment]
Pull requests opened to fix D300, D200 and D204 in Tests - a bit more effort on this and we can have a single ``.flake8`` configuration, and use the default flake8 git pre-commit hook - which will be simpler to document and maintain 👍

**Comment by peterjc** [comment]
I propose to solve E731 the pragmatic way - whitelist most of them with ``#noqa`` lines, see #2019.

I can fix D100 mostly automatically with the following script (missing docstrings can trigger false edits, and it leaves question marks and exclamation marks to a human to edit), will do this once the current pull requests are merged:

```python
import sys


def fix(lines):
    answer = []
    post_def = False
    for line in lines:
        if line.lstrip().startswith("def "):
            # print(line.rstrip())
            post_def = True
        if post_def and line.lstrip().startswith('"""'):
            post_def = False
            # Looks like line one of a docstring
            if line.rstrip().endswith('"""'):
                # a one-liner
                if not line.rstrip().endswith(('."""', '?"""', '!"""')):
                    print(line.rstrip())
                    line = line.rstrip()[:-3] + '."""\n'
            else:
                # multi-liner
                if not line.rstrip().endswith(('.', '?', '!')):
                    print(line.rstrip())
                    line = line.rstrip() + ".\n"
        answer.append(line)
    return answer


for f in sys.argv[1:]:
    print(f)
    with open(f) as handle:
        old = list(handle)
    new = fix(old)
    with open(f, "w") as handle:
        for line in new:
            handle.write(line)
print("Done")
```

Next would be ``D401 First line should be in imperative mood`` and there are now few enough of these to do my hand later (again, waiting on the open pull requests as I already fixed some).

**Comment by MarkusPiotrowski** [comment]
@peterjc Do you know that newer ```flake8``` versions (since 3.7) have a ```per-file-ignores``` option? This allows to set up only one ```.flake8``` file and still have folder specific ignores, e.g.:

```
[flake8]
doctests = True
ignore =
    E122,E123,E126,W503,W504,
    E501,E731,F401,F841,
    D101,D102,D103,D105,
    D203,
    B007,B009,B010,B011,
    C812,C815,
    BLK100,
per-file-ignores =
    /Bio/*:W303
```
This example doesn't make too much sense, I just used it for testing: I put too much empty lines in one file in ```Bio``` and one file in ```Tests```, and indeed, only the file from the Tests folder was flagged.

**Comment by peterjc** [comment]
@MarkusPiotrowski I did not know that - if as in your example it works nicely with folder level settings, that would do nicely (not that fixing the issues in ``Tests/`` isn't a good idea anyway).

Once we've merged #2019 and #2020, do you want to try a single strict top-level ``.flake8`` instead of #2018 - taking care of E731 etc as extra ignores for ``Bio/``, and D202 etc as extra ignores on the ``Tests/`` folder?

**Comment by MarkusPiotrowski** [comment]
I found this just today. Didn't want that you spend your whole Easter Holidays with these issues...
Yes, I would like to give it a try. Do you need to update flake8?

**Comment by peterjc** [comment]
One of the interesting things about my employer is we don't get public holidays (although the institute does close over Christmas into New Year), so if we take them off they come out of (reasonably generous) annual leave. I'll be back at work on Monday.

--

By all means, please have a play with the combined settings - I'm not planning on doing any more flake8 work today or the weekend (aside perhaps from merging pull requests).

**Comment by peterjc** [comment]
I think we're now very close to having ``Bio/`` and ``Tests/`` having the same ``.flake8`` settings, with ``BioSQL/`` and ``Doc/examples/`` even stricter. So will be soon be able to use a single top level ``.flake8`` file.

The quick solution will be to adopt the ``Bio/.flake8`` as the sole top level settings - but ideally using Markus' idea lets us have the stricter settings with additional ignores under ``Bio/`` and ``Tests/`` as needed 👍

**Comment by MarkusPiotrowski** [comment]
I'm actually working on style issues in ```Scripts```.
I already have made a 'top-level' ```.flake8``` and I'm waiting now for some of the ongoing style-related PR's to be merged to test it.

**Comment by peterjc** [comment]
This time round, let's put each ignore code on its own line (sensibly sorted) as it will make tracking changes much easier?

**Comment by MarkusPiotrowski** [comment]
Unfortunately, it seems that the ```per-file-ignores``` must be all on the same line per file pattern. I tried to put them one on each line with following comments, but that didn't work.

**Comment by MarkusPiotrowski** [comment]
With this ```.flake8``` file (not up-to-date regarding the running PRs):
```
[flake8]
doctests = True
ignore =
    # =============================================================
    # Biopython's 'standard' ignores we can agree to always accept:
    # =============================================================
    E501, # line too long TODO: Maybe we find a sensible limit, e.g. 88 (black) and enforce it
    W503, # line-break before binary operator, deliberately ignore in favour of W504
    D203, # 1 blank line required before class docstring
          # We ignore D203 deliberately in favour of passing D211 No blank lines allowed before
          # class docstring
    # ===========================================
    # Ignores that we have to accept for a while:
    # ===========================================
    E123, # closing bracket does not match indentation of opening bracket's line (top, Bio, Tests, 132 occurrences)
    W504, # line break after binary operator TODO?
    B007, # Loop control variable not used within the loop body.
          # If this is intended, start the name with an underscore.
    #==========================================
    # Optional ignores for local installations:
    # =========================================
    BLK100, # Black would make changes, only on local installations (so far)
per-file-ignores =
    Bio/*:E122,E126,E731,F401,F841,D101,D102,D103,D105,B009,B010,B011,C812,C815
    Tests/*:E122,E126,E241,E401,E731,F401,F405,F841,D100,D101,D102,D103,D202,D400,D401,D412,B009,B010,B011,C812

# This format is not allowed:
# per-file-ignores =
#    /Bio/*:E122, # continuation line missing indentation or outdented TODO (271 occurrences)
#           E126, # continuation line over-indented for hanging indent TODO  (53 occurrences)
#           E731, # do not assign a lambda expression, use a def TODO (28 occurrences)
#           F401, # module imported but unused TODO (108 occurrences)
#           F841, # local variable is assigned to but never used TODO (55 occurrences)
#           D101, # missing docstring in public class TODO (43 occurrences)
#           D102, # missing docstring in public method TODO (81 occurrences)
#           D103, # missing docstring in public function TODO (17 occurrences)
#           D105, # missing docstring magic method TODO (116 occurrences)
#           B009, # do not call getattr with a constant attribute value,
#                 # it is not any safer than normal property access
#           B010, # do not call setattr with a constant attribute value,
#                 # it is not any safer than normal property access
#           B011, # Do not call assert False since python -O removes these calls;
#                 # instead callers should raise AssertionError().
#           C812, # missing trailing comma
#          C815  # missing trailing comma in Python 3.5+
#    /Tests/*:E122, # continuation line missing indentation or outdented TODO (48 occurrences)
#             E126, # continuation line over-indented for hanging indent TODO (41 occurrences)
#             E241, # multiple spaces after ',' TODO (54 occurrences)
#             E401, # module level import not at top of file TODO? (and noqa?) (49 occurrences)
#             E731, # do not assign a lambda expression, use a def, we want deliberately ignore this in tests
#             F401, # module imported but unused TODO (76 occurrences)
#             F405, # name may be undefined, or defined from star imports: module TODO or noqa?,only 21 in tests
#             F841, # local variable is assigned to but never used TODO? (68 occurrences)
#             D100, # missing docstring in public module, TODO (75 occurrences)
#             D101, # missing docstring in public class (220 occurrences)
#             D102, # missing docstring in public method (976 occurrences)
#             D103, # missing docstring in public functions TODO (50 occurrences)
#             D202, # no blank lines allowed after function docstring (88 occurrences)
#             D400, # first line should end with a period (742 occurrences)
#             D401, # first line should be in imperative mood TODO(31 occurrences)
#             D412, # no blank lines allowed between a section header and its content TODO(1 occurrence)
#             B009, # do not call getattr with a constant attribute value,
#                   # it is not any safer than normal property access
#             B010, # do not call setattr with a constant attribute value,
#                   # it is not any safer than normal property access
#             B011, # Do not call assert False since python -O removes these calls;
#                   # instead callers should raise AssertionError()
#             C812  # missing trailing comma
``` 
we get the following result:
```
G:\git_32bit\biopython>flake8 --statistics --quiet
.\setup.py
.\Doc\examples\ACT_example.py
.\Doc\examples\fasta_iterator.py
.\Doc\examples\fasta_dictionary.py
.\Doc\examples\getgene.py
.\Scripts\query_pubmed.py
.\Scripts\scop_pdb.py
.\Scripts\debug\debug_blast_parser.py
.\Scripts\GenBank\check_output.py
.\Scripts\Restriction\ranacompiler.py
.\Scripts\Restriction\rebase_update.py
.\Scripts\xbbtools\nextorf.py
.\Scripts\xbbtools\xbb_blast.py
.\Scripts\xbbtools\xbb_blastbg.py
.\Scripts\xbbtools\xbb_help.py
.\Scripts\xbbtools\xbb_search.py
.\Scripts\xbbtools\xbb_translations.py
.\Scripts\xbbtools\xbb_utils.py
.\Scripts\xbbtools\xbb_widget.py
.\Tests\test_NaiveBayes.py
.\Tests\test_seq.py
.\Tests\test_PDB_ResidueDepth.py
.\Tests\test_ExPASy.py
.\Tests\test_KGML_graphics_online.py
.\Tests\test_SeqIO_online.py
.\Tests\test_PDBList.py
.\Tests\test_TogoWS.py
.\Tests\test_phyml_tool.py
.\Tests\test_Uniprot.py
.\Tests\test_psw.py
.\Tests\test_SCOP_online.py
.\Tests\requires_wise.py
.\Tests\test_Wise.py
.\Tests\test_ClustalOmega_tool.py
.\Tests\test_Entrez_parser.py
.\Tests\test_HMMCasino.py
1     D100 Missing docstring in public module
18    D101 Missing docstring in public class
95    D102 Missing docstring in public method
14    D103 Missing docstring in public function
2     D105 Missing docstring in magic method
15    D107 Missing docstring in __init__
1     E126 continuation line over-indented for hanging indent
49    E402 module level import not at top of file
```
This does (of course) not include the running PRs, e.g. all the issues in ```/Scripts``` and ```/Tests``` are probably already solved and there may be only some single files that need further attention.

**Comment by peterjc** [comment]
Do you know if this style would work? If so could opt for  even more path repetition to get one code per line?

```
per-file-ignores =
    Bio/*:E122,E126,E731,
    Bio/*:F401,F841,
    Bio/*:D101,D102,D103,D105,
    Bio/*:B009,B010,B011,
    Bio/*:C812,C815
    Tests/*:E122,E126,E241,E401,E731,
    Tests/*:F401,F405,F841,
    Tests/*:D100,D101,D102,D103,D202,D400,D401,D412,
    Tests/*:B009,B010,B011,C812
```

(If you don't known, I might try this later today)

**Comment by MarkusPiotrowski** [comment]
I had the same idea, but haven't tested yet. So, no I don't know.

**Comment by MarkusPiotrowski** [comment]
It does seem to work. 

However, I have a different issue, which would make the single-line version more attractive:
When running `flake8` from command line, the path in the `per-file-ignores` must be:
```
Bio/*:E122,...
```
However, my pre-commit hook does only work correctly when the path is given with a starting backslash:
```
/Bio/*:E122,...
```
Can you confirm this?

**Comment by peterjc** [comment]
Are you using the ``flake8`` provided hook?

```
$ flake8 --install-hook git
$ git config --bool flake8.strict true
```

**Comment by MarkusPiotrowski** [comment]
Yes, I think so, everything in place (`biophython/.git/hooks/pre-commit`).

**Comment by MarkusPiotrowski** [comment]
I wonder if this is related to a [known issue](https://gitlab.com/pycqa/flake8/issues/517) with `per-file-ignores`:
Test: I added some style violations into `Bio/_utils.py` :
```
PS G:\git_32bit\Biopython> flake8 Bio/_utils.py --isolated
Bio/_utils.py:1:1: D100 Missing docstring in public module
Bio/_utils.py:8:1: E265 block comment should start with '# '
Bio/_utils.py:41:1: E303 too many blank lines (4)
Bio/_utils.py:45:80: E501 line too long (114 > 79 characters)
```
I added `D100` to the `per-file-ignores` for `Bio\*` and ran the test with the `.flake8` file:
```
PS G:\git_32bit\Biopython> flake8 Bio/_utils.py
Bio/_utils.py:8:1: E265 block comment should start with '# '
Bio/_utils.py:41:1: E303 too many blank lines (4)
```
Everything's fine, E501 is ignored via `ignores` and D100 is ignored via `per-file-ignores`. Now, I go one directory up:
```
PS G:\git_32bit\Biopython> cd..
PS G:\git_32bit> flake8 Biopython/Bio/_utils.py
Biopython/Bio/_utils.py:1:1: D100 Missing docstring in public module
Biopython/Bio/_utils.py:8:1: E265 block comment should start with '# '
Biopython/Bio/_utils.py:41:1: E303 too many blank lines (4)
```
E501 is still ignored, indicating the `.flake8` file is used, but the `per-file-ignores` (D100) don't work anymore.

**Comment by peterjc** [comment]
Could be https://gitlab.com/pycqa/flake8/issues/517 ?

**Comment by peterjc** [comment]
I think you could be right, https://gitlab.com/pycqa/flake8/issues/517 looks likely to be the problem.

**Comment by MarkusPiotrowski** [comment]
It would be nice if someone can confirm that a `.flake8` file within the Biopython folder with `per-file-ignores` like this:
```
[flake8]
ignore =E501,W503,D203,E123,W504,B007,BLK100,D100,E126
per-file-ignores =
    Bio/*:E122,E731,F401,F841,D101,D102,D103,D105,B009,B010,B011,C812,C815
    Tests/*:E122,E731,F401,F841,D101,D102,D103,B009,B010,B011,C812
    Scripts/*:D10
```
works in principal when `flake8` is called from the command line within the Biopython folder (`..biopython/>flake8 --statistics --quiet --quiet`) but not when it is used by a pre-commit hook when doing a `git commit` (e.g. put an unused import in one file in `Bio` or `Tests` ->F401). When changing the globs to:
```
per-file-ignores =
    /Bio/*:E122,E731,F401,F841,D101,D102,D103,D105,B009,B010,B011,C812,C815
    /Tests/*:E122,E731,F401,F841,D101,D102,D103,B009,B010,B011,C812
    /Scripts/*:D10
```
it's *vice versa*. Once this is confirmed, we can file an issue to the `flake8` people.

In the meanwhile, I have found a workaround which works for me, which is simply adding both paths:
```
per-file-ignores =
    Bio/*:E122,E126,E731,F401,F841,D101,D102,D103,D105,B009,B010,B011,C812,C815
    /Bio/*:E122,E126,E731,F401,F841,D101,D102,D103,D105,B009,B010,B011,C812,C815
    Tests/*:E122,E126,E731,F401,F841,D101,D102,D103,B009,B010,B011,C812
    /Tests/*:E122,E126,E731,F401,F841,D101,D102,D103,B009,B010,B011,C812
```

**Comment by MarkusPiotrowski** [comment]
With the last commits and a single `.flake8` file that looks principially like this (removed comments etc):
```
[flake8]
doctests = True
ignore =E501,W503,D203,E123,W504,B007,BLK100
per-file-ignores =
    Bio/*:E122,E126,E731,F401,F841,D101,D102,D103,D105,B009,B010,B011,C812,C815
    Tests/*:E122,E126,E731,F401,F841,D101,D102,D103,B009,B010,B011,C812
```
we have only 5 files left that need some attention:
```
H:\git_32bit\biopython>flake8 --statistics
.\setup.py:391:17: E126 continuation line over-indented for hanging indent
.\Doc\examples\ACT_example.py:1:1: D100 Missing docstring in public module
.\Doc\examples\fasta_iterator.py:13:1: D103 Missing docstring in public function

.\Doc\examples\fasta_dictionary.py:30:1: D103 Missing docstring in public functi
on
.\Doc\examples\getgene.py:49:1: D101 Missing docstring in public class
.\Doc\examples\getgene.py:58:1: D102 Missing docstring in public method
.\Doc\examples\getgene.py:92:1: D102 Missing docstring in public method
.\Doc\examples\getgene.py:100:1: D102 Missing docstring in public method
.\Doc\examples\getgene.py:103:1: D102 Missing docstring in public method
.\Doc\examples\getgene.py:113:1: D102 Missing docstring in public method
.\Doc\examples\getgene.py:128:1: D102 Missing docstring in public method
.\Doc\examples\getgene.py:133:1: D102 Missing docstring in public method
.\Doc\examples\getgene.py:147:1: D102 Missing docstring in public method
.\Doc\examples\getgene.py:165:1: D102 Missing docstring in public method
.\Doc\examples\getgene.py:180:1: D102 Missing docstring in public method
.\Doc\examples\getgene.py:202:1: D102 Missing docstring in public method
.\Doc\examples\getgene.py:224:1: D102 Missing docstring in public method
.\Doc\examples\getgene.py:243:1: D102 Missing docstring in public method
.\Doc\examples\getgene.py:257:1: D102 Missing docstring in public method
.\Doc\examples\getgene.py:276:1: D103 Missing docstring in public function
1     D100 Missing docstring in public module
1     D101 Missing docstring in public class
14    D102 Missing docstring in public method
3     D103 Missing docstring in public function
1     E126 continuation line over-indented for hanging indent
```

**Comment by MarkusPiotrowski** [comment]
All gone!
```
H:\git_32bit\biopython>flake8 --statistics --quiet

H:\git_32bit\biopython>
```


Regarding the format of the `per-file-ignores` line (I'm not speaking about the path issue now), it seems that your idea, i.e. to add several `Bio/*:` lines with different error codes does *not* seem to work (in my first test I took an unfortunate example so I was mislead that this works). Obviously, for one _glob_ only one error 'line' is allowed; if there are several identical globs, one will be selected ('the last match sorted alphabetically'). 

However, what can be done is to separate the error codes by *whitespaces* (e.g. line breaks!). Unfortunately, one cannot put an error code with a comment on the same line, but on different lines. So this would look like so:
```
per-file-ignores =
    Bio/*:E122
          # continuation line missing indentation or outdented TODO (268 occurrences)
          E126
          # continuation line over-indented for hanging indent TODO  (53 occurrences)
          E731
          # do not assign a lambda expression, use a def TODO (28 occurrences)
          ...
```

There would be many of these and I can not judge if this is clearer or easier to read than a format like this:
``` 
per-file-ignores =         
   Bio/*:E122,E126,E731,F401,F841,D101,D102,D103,D105,B009,B010,B011,C812,C815
   Tests/*:E122,E126,E731,F401,F841,D101,D102,D103,B009,B010,B011,C812

# per-file-ignores =
# Bio/*:E122, # continuation line missing indentation or outdented TODO (268 occurrences)
#       E126, # continuation line over-indented for hanging indent TODO  (53 occurrences)
#       E731, # do not assign a lambda expression, use a def TODO (28 occurrences)
...
```

Any opinions about this?

**Comment by peterjc** [comment]
Good work on ``Doc/examples/``  👍 

Without the flake8 bug/workaround, I would definitely prefer the first configuration version (allowing one line per style code interspaced with explanatory comments). This will behave best with edits and minimise merge conflicts.

However, given it seems we must list all the codes twice as a flake8 workaround, two-multi-line lists would be very verbose and harder to keep in sync, so two 'all codes in one long line' (and a shared comment block) seems the least bad solution.

Sorry, I've not get tried to reproduce the flake8 configuration issue.

**Comment by MarkusPiotrowski** [comment]
@peterjc I filed an issue to the flake8 people. Maybe you can have look at the discussion (https://gitlab.com/pycqa/flake8/issues/538), since they suggest some optimizations by using `pre-commit.com`

**Comment by peterjc** [comment]
https://gitlab.com/pycqa/flake8/issues/538#note_169613825 is interesting, it sounds like flake8 may well start to recommend using ``pre-commit`` over the current ``flake8`` provided hook.

**Comment by MarkusPiotrowski** [comment]
In their docs they are already pretty clear about this: http://flake8.pycqa.org/en/latest/user/using-hooks.html

**Comment by peterjc** [comment]
Added 21 Nov 2018, I think I'd missed that.

I agree that moving to ``pre-commit`` would allow including the misc checks currently done on TravisCI via the Tox configuration too - so that would be a practical improvement.

I suspect we could even use our current multiple ``.flake8`` file setup with ``pre-commit``, but I would rather move to a single top level ``.flake8`` instead.

**Comment by peterjc** [comment]
With #2074 merged, @MarkusPiotrowski are you happy for us to recommend the flake8 pre-commit hook in the ``CONTRIBUTING.rst`` file (etc)?

**Comment by peterjc** [comment]
Good progress, I've closed #493 as of #2075.

Still need to mention the hook on https://biopython.org/wiki/GitUsage

Also consider if and how to mention it in the ``PULL_REQUEST_TEMPLATE.md`` too.

**Comment by peterjc** [comment]
How about this:

```
- [ ] I have read the ``CONTRIBUTING.rst`` file, have run ``flake8`` locally, and
understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.
```

The wording aims to cover the fact that TravisCI/Tox includes some extra style checks not currently done via flake8 (e.g. the RST linting)

**Comment by MarkusPiotrowski** [comment]
Fine!

**Comment by peterjc** [comment]
Change applied - are we OK to close this issue?

**Comment by MarkusPiotrowski** [comment]
Yes, I'm doing so now. Thank you!

**Comment by peterjc** [comment]
Excellent work, thank you too.
